File: clients/python/src/agenticdriver/panel.py
"""Shared provider component and backend bridge; mount behind the application's settings authorization."""
import base64
import html
import re
from importlib.resources import files
from typing import Any, Callable, Awaitable
from .client import AgenticClient
from .async_client import AsyncAgenticClient
from ._errors import DriverError

def provider_panel_script() -> bytes:
    return files("agenticdriver").joinpath("static/provider-panel.js").read_bytes()

def provider_panel_html(api_path: str = "/api/agenticdriver-panel", module_path: str = "/assets/agenticdriver-panel.js") -> str:
    for path in [api_path, module_path]:
        if not re.fullmatch(r"/(?!/)[^\s?#\\]*", path) or ".." in path.split("/"):
            raise ValueError("Use same-origin absolute asset and API paths.")
    return f'<agenticdriver-providers api="{html.escape(api_path, quote=True)}"></agenticdriver-providers><script type="module" src="{html.escape(module_path, quote=True)}"></script>'

class ProviderPanel:
    """Resolve an already connected SDK client; optional hooks own private connection persistence."""
    def __init__(self, client: Callable[[], AgenticClient | None], *, connection: Callable[[], dict[str, str] | None] | None = None, connect: Callable[[str], None] | None = None, disconnect: Callable[[], None] | None = None):
        self.client = client
        self.connection = connection or (lambda: None)
        self.connect = connect
        self.disconnect = disconnect

    def snapshot(self, refresh: bool = False) -> dict[str, Any]:
        client = self.client()
        state: dict[str, Any] = {"connected": client is not None, "providers": [], "canConnect": self.connect is not None, "canDisconnect": self.disconnect is not None, "canInvite": False}
        if self.connection(): state["connection"] = self.connection()
        if client is None: return state
        state["providers"] = client.providers(refresh=refresh)
        features = client.protocol()["features"]
        if "provider-management" in features:
            state["management"] = client.management()
            state["canInvite"] = bool(state.get("connection", {}).get("url") and "client-pairing" in features)
        return state

    def handle(self, request: dict[str, Any]) -> Any:
        action = request.get("action")
        if action == "snapshot": return self.snapshot(request.get("refresh") is True)
        if action == "connect" and self.connect:
            invitation = request.get("invitation")
            if not isinstance(invitation, str) or len(invitation) > 16384: raise DriverError("INVALID_INVITATION", "Use a complete host invitation.")
            self.connect(invitation)
            return self.snapshot()
        if action == "disconnect" and self.disconnect:
            self.disconnect()
            return self.snapshot()
        client = self.client()
        if client is None: raise DriverError("CONNECTION_REQUIRED", "Connect an AgenticDriver host first.")
        if action == "configure":
            client.configure_provider(request["change"])
            return self.snapshot()
        if action == "connections": return client.connections()
        if action == "revoke": return {"revoked": client.revoke_connection(request["id"])}
        if action == "invite":
            url = (self.connection() or {}).get("url")
            if not url: raise DriverError("CONNECTION_UNAVAILABLE", "Configure the host's public connection URL.")
            invite = client.create_invitation({"grant": {"subject": request["subject"], "providers": request["providers"], "manageProviders": request.get("manageProviders") is True}})
            encoded = base64.urlsafe_b64encode(url.encode("ascii")).decode().rstrip("=")
            return {"invitation": f'ad1.{encoded}.{invite["code"]}', "expiresAt": invite["expiresAt"], "grant": invite["grant"]}
        raise DriverError("INVALID_PANEL_REQUEST", "Choose a supported provider panel operation.")

class AsyncProviderPanel:
    """Same component bridge with async SDK client and async connection hooks."""
    def __init__(self, client: Callable[[], AsyncAgenticClient | None], *, connection: Callable[[], dict[str, str] | None] | None = None, connect: Callable[[str], Awaitable[None]] | None = None, disconnect: Callable[[], Awaitable[None]] | None = None):
        self.client = client
        self.connection = connection or (lambda: None)
        self.connect = connect
        self.disconnect = disconnect

    async def snapshot(self, refresh: bool = False) -> dict[str, Any]:
        client = self.client()
        state: dict[str, Any] = {"connected": client is not None, "providers": [], "canConnect": self.connect is not None, "canDisconnect": self.disconnect is not None, "canInvite": False}
        if self.connection(): state["connection"] = self.connection()
        if client is None: return state
        state["providers"] = await client.providers(refresh=refresh)
        features = (await client.protocol())["features"]
        if "provider-management" in features:
            state["management"] = await client.management()
            state["canInvite"] = bool(state.get("connection", {}).get("url") and "client-pairing" in features)
        return state

    async def handle(self, request: dict[str, Any]) -> Any:
        action = request.get("action")
        if action == "snapshot": return await self.snapshot(request.get("refresh") is True)
        if action == "connect" and self.connect:
            invitation = request.get("invitation")
            if not isinstance(invitation, str) or len(invitation) > 16384: raise DriverError("INVALID_INVITATION", "Use a complete host invitation.")
            await self.connect(invitation)
            return await self.snapshot()
        if action == "disconnect" and self.disconnect:
            await self.disconnect()
            return await self.snapshot()
        client = self.client()
        if client is None: raise DriverError("CONNECTION_REQUIRED", "Connect an AgenticDriver host first.")
        if action == "configure":
            await client.configure_provider(request["change"])
            return await self.snapshot()
        if action == "connections": return await client.connections()
        if action == "revoke": return {"revoked": await client.revoke_connection(request["id"])}
        if action == "invite":
            url = (self.connection() or {}).get("url")
            if not url: raise DriverError("CONNECTION_UNAVAILABLE", "Configure the host's public connection URL.")
            invite = await client.create_invitation({"grant": {"subject": request["subject"], "providers": request["providers"], "manageProviders": request.get("manageProviders") is True}})
            encoded = base64.urlsafe_b64encode(url.encode("ascii")).decode().rstrip("=")
            return {"invitation": f'ad1.{encoded}.{invite["code"]}', "expiresAt": invite["expiresAt"], "grant": invite["grant"]}
        raise DriverError("INVALID_PANEL_REQUEST", "Choose a supported provider panel operation.")

File: src/panel.ts
import { z } from "zod";
import { AgenticClient, connectionInvitation } from "./client.js";
import {
  ConfigureProviderSchema,
  type ManagementSnapshot,
} from "./management-types.js";
import { DriverError } from "./errors.js";
import {
  ProviderSetupRequestSchema,
  type ProviderSetupSnapshot,
} from "./setup-types.js";
import type { ProviderInfo } from "./types.js";
import type { ProviderPresentation } from "./catalog.js";
export interface PanelConnection {
  id: string;
  label: string;
  url?: string;
}
export interface ProviderPanelState {
  connected: boolean;
  connection?: PanelConnection;
  providers: ProviderInfo[];
  management?: ManagementSnapshot;
  setup?: ProviderSetupSnapshot;
  presentations?: Record<string, ProviderPresentation>;
  canConnect: boolean;
  canDisconnect: boolean;
  canInvite: boolean;
  connectionError?: string;
}
export const PanelRequestSchema = z.discriminatedUnion("action", [
  z
    .object({ action: z.literal("setup"), request: ProviderSetupRequestSchema })
    .strict(),
  z
    .object({ action: z.literal("snapshot"), refresh: z.boolean().optional() })
    .strict(),
  z
    .object({ action: z.literal("configure"), change: ConfigureProviderSchema })
    .strict(),
  z
    .object({
      action: z.literal("connect"),
      invitation: z.string().min(1).max(16_384),
    })
    .strict(),
  z.object({ action: z.literal("disconnect") }).strict(),
  z
    .object({
      action: z.literal("invite"),
      subject: z.string().min(1).max(128),
      providers: z.array(z.string()).max(32),
      manageProviders: z.boolean().default(false),
    })
    .strict(),
  z.object({ action: z.literal("connections") }).strict(),
  z.object({ action: z.literal("revoke"), id: z.uuid() }).strict(),
]);
export type PanelRequest = z.input<typeof PanelRequestSchema>;
export interface ProviderPanelOptions {
  /** Resolve the currently connected backend client for the authorized application user. */
  client(): AgenticClient | undefined | Promise<AgenticClient | undefined>;
  connection?(): PanelConnection | undefined;
  connect?(invitation: string): Promise<void>;
  disconnect?(): Promise<void>;
  /** Reuse reviewed Usagestat metadata/assets through providerPresentation(). */
  presentations?(
    providers: ProviderInfo[],
  ):
    | Record<string, ProviderPresentation>
    | Promise<Record<string, ProviderPresentation>>;
}
/** Mount behind the consuming application's existing settings authorization and CSRF checks. */
export function providerPanel(options: ProviderPanelOptions) {
  async function snapshot(refresh = false): Promise<ProviderPanelState> {
    const client = await options.client();
    const base = {
      canConnect: Boolean(options.connect),
      canDisconnect: Boolean(options.disconnect),
      canInvite: false,
    };
    if (!client)
      return {
        ...base,
        connected: false,
        connection: options.connection?.(),
        providers: [],
      };
    const providers = await client.providers({ refresh });
    let management: ManagementSnapshot | undefined;
    const protocol = await client.protocol();
    if (protocol.features.includes("provider-management"))
      management = await client.management();
    const connection = options.connection?.();
    return {
      ...base,
      connected: true,
      providers,
      connection,
      management,
      ...(management && protocol.features.includes("provider-setup")
        ? { setup: await client.providerSetup({ action: "list" }) }
        : {}),
      canInvite: Boolean(
        management &&
        connection?.url &&
        protocol.features.includes("client-pairing"),
      ),
      ...(options.presentations
        ? { presentations: await options.presentations(providers) }
        : {}),
    };
  }
  return async (input: unknown): Promise<unknown> => {
    const request = PanelRequestSchema.safeParse(input);
    if (!request.success)
      throw new DriverError(
        "INVALID_PANEL_REQUEST",
        "Choose a supported provider panel operation.",
      );
    const action = request.data;
    if (action.action === "snapshot") return snapshot(action.refresh);
    if (action.action === "connect") {
      if (!options.connect)
        throw new DriverError(
          "CONNECTION_UNAVAILABLE",
          "This application has not enabled connection setup.",
        );
      await options.connect(action.invitation);
      return snapshot();
    }
    if (action.action === "disconnect") {
      if (!options.disconnect)
        throw new DriverError(
          "CONNECTION_UNAVAILABLE",
          "This application manages its connection outside this panel.",
        );
      await options.disconnect();
      return snapshot();
    }
    const client = await options.client();
    if (!client)
      throw new DriverError(
        "CONNECTION_REQUIRED",
        "Connect an AgenticDriver host first.",
      );
    if (action.action === "configure") {
      await client.configureProvider(action.change);
      return snapshot();
    }
    if (action.action === "setup") return client.providerSetup(action.request);
    if (action.action === "connections") return client.connections();
    if (action.action === "revoke") return client.revokeConnection(action.id);
    const connection = options.connection?.();
    if (!connection?.url)
      throw new DriverError(
        "CONNECTION_UNAVAILABLE",
        "The host's public connection URL is not configured.",
      );
    const invite = await client.createInvitation({
      grant: {
        subject: action.subject,
        providers: action.providers,
        manageProviders: action.manageProviders,
      },
    });
    return {
      invitation: connectionInvitation(connection.url, invite.code),
      expiresAt: invite.expiresAt,
      grant: invite.grant,
    };
  };
}
